### DISCUSSION TITLE
Native BigInt in Javascript

### DISCUSSION THREAD

**Discussion by ericman314** [josdejong/mathjs#2737] [Design decisions]
Chrome 67 will ship with [native support](https://developers.google.com/web/updates/2018/05/bigint) for the `BigInt` type, arbitrary precision itegers. I opened this issue to discuss how we might use `BigInt`s once adoption is more widespread.

There's no `BigDecimal` implementation yet, but `BigInt` could be particularly useful for fractions.

Related: #15, #604

**Comment by ericman314** [comment]
There could also be the possibility of implementing a very simple BigDecimal library which uses BigInts under the hood for exceptional performance. I will look into this!

**Comment by josdejong** [comment]
Yes, really cool!

Maybe we should simply introduce this as a separate, new data type `BigInt` (alongside the existing `BigDecimal`). 

Having a `BigInt` will be really interesting to use in Fractions. That should become relatively easy when we have implemented #694 for Complex numbers and Fractions (Unit already has supports multiple numeric types).

**Comment by ericman314** [comment]
I've published a small library that implements big decimal numbers using the new `BigInt` type. It currently works in Chrome 67 and Node.js 10. It's not much, but it could come in handy.

https://github.com/ericman314/bignum-native

If anyone thinks it might have promise I would sure appreciate any contributions you can make.

**Comment by josdejong** [comment]
Thanks for sharing Eric!

**Comment by harrysarson** [comment]
It would be great if somehow users could pick their big number implementation.

**Comment by josdejong** [comment]
Here is a small proof of concept of using BigInt in mathjs:

http://jsbin.com/kikokot/edit?html,output

```js
<!DOCTYPE html>
<html>
<head>
  <meta name="description" content="math.js | bigint demo">
  <title>math.js | bigint demo</title>
  <script src="https://unpkg.com/mathjs@5.1.2/dist/math.min.js"></script>
</head>
<body>
  <h1>mathjs with BigInt demo</h1>
  <p>
    Only works on the latest version of Chrome right now
  </p>
  
  <p id="addResult"></p>
  <p id="powResult"></p>
  <p id="error" style="color: red"></p>
  
  <script>
    try {
      math.import({
        name: 'BigInt',
        path: 'type', // will be imported into math.type.BigInt
        factory: (type, config, load, typed) => {
          typed.addType({
            name: 'BigInt',
            test: (x)  => typeof x === 'bigint'
          })

          // we can also add conversions here from number or string to BigInt
          // and vice versa using typed.addConversion(...)

          return BigInt
        },
        lazy: false // disable lazy loading as this factory has side
                    // effects: it adds a type and a conversion.
      })

      math.import({
        name: 'bigint',
        factory: (type, config, load, typed) => {
          return typed('bigint', {
            'number | string ': (x) => BigInt(x)
          })
        }
      })

      math.import({
        name: 'add',
        factory: (type, config, load, typed) => {
          return typed('add', {
            'BigInt, BigInt': (a, b)  => a + b
          })
        }
      })

      math.import({
        name: 'pow',
        factory: (type, config, load, typed) => {
          return typed('pow', {
            'BigInt, BigInt': (a, b)  => a ** b
          })
        }
      })

      document.getElementById('addResult').innerHTML = 
        '4349 + 5249 = ' + math.eval('bigint(4349) + bigint(5249)')

      document.getElementById('powResult').innerHTML = 
        '4349 ^ 5249 = ' + math.eval('bigint(4349) ^ bigint(5249)')
    }
    catch (err) {
      console.error(err)
      document.getElementById('error').innerHTML = err.toString()
    }
  </script>
</body>
</html>
```

**Comment by tmtron** [comment]
Are there any news/plans for BigInt support?  
As of 2022 all major browsers have [BigInt support](https://caniuse.com/bigint)?

**Reply by josdejong** [reply]
Yes definitely, it will be a great addition. Support for `BigInt` is one of the reasons that we're working on a large architectural overhaul of the library. We could implement it `BigInt` right now, but the current architecture is not ideal and we would like to seize the opportunity to refactor the library, after which adding `BigInt` support will be simpler and more flexible. 

At this moment @gwhitney is working out a POC for the new architecture, which is looking very promising!

You can read more in https://github.com/josdejong/mathjs/issues/2800 (I've just explicitly added a bullet point for `BigInt` support there)

**Reply by gwhitney** [reply]
As Jos bit the buller and folded bigints into the current architecture, closing this discussion.
